Fixed two System.Xml uap errors

ViktorHofer opened the pull request:
Basically moving to temporary paths and using the right overload in XmlReader. The hard thing was to find out that xml validation will fail if the directory doesn't have a trailing slash.

Thanks krwq!

https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fuwp~2F/build/20170512.03/workItem/System.Xml.XmlSchema.XmlSchemaValidatorApi.Tests/analysis/xunit/System.Xml.Tests.TCXmlSchemaValidatorMisc~2FXSDValidationGeneratesInvalidError_1

https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fuwp~2F/build/20170512.03/workItem/System.Xml.XmlSchemaSet.Tests/analysis/xunit/System.Xml.Tests.TC_SchemaSet_Compile~2FTFS_470021

danmosemsft commented on src/System.Private.Xml/tests/XmlSchema/XmlSchemaSet/TC_SchemaSet_Compile.cs:
```diff
@@ -118,7 +118,12 @@ public void TFS_470021()
 </xs:schema>";
-            using (XmlWriter w = XmlWriter.Create("cham.xsd"))
+
+            string testDirectory = Path.Combine(Path.GetTempPath(), GetType().Name + "_" + Path.GetRandomFileName() + Path.DirectorySeparatorChar);
```
Please reuse TempDirectory/TempFile classes instead

danmosemsft commented on src/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ValidateMisc.cs:
```diff
@@ -912,19 +916,26 @@ public void XSDValidationGeneratesInvalidError_1()
         [Fact]
         public void XSDValidationGeneratesInvalidError_2()
         {
+            // TestDirectory path must end with a DirectorySaratorChar, otherwise it will throw in the Xml validation.
```
Typo

danmosemsft commented on src/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ValidateMisc.cs:
```diff
@@ -935,7 +946,10 @@ public void XSDValidationGeneratesInvalidError_2()
-            return;
+
+            try
+            { Directory.Delete(testDirectory, recursive: true); }
```
If you use existing temp file helpers, using statement will take care of this

ViktorHofer commented:
It seems there is a different behavior in Validation on Unix/OSX vs Windows. Need to investigate...

krwq commented:
I think you might be missing baseURI param in the failing tests - for the appending slash - create a function called EnsureTrailingSlash and make sure you do not have double slash in the end (TempDirectory might be returning already with a slash on linux)

ViktorHofer commented:
Could be... Let's see :)

krwq commented on src/System.Private.Xml/tests/XmlSchema/XmlSchemaSet/TC_SchemaSet_Compile.cs:
```diff
@@ -118,33 +118,42 @@ public void TFS_470021()
-            XmlSchemaSet ss = new XmlSchemaSet();
-            ss.ValidationEventHandler += new ValidationEventHandler(ValidationCallback);
+                string chamPath = Path.Combine(tempDirectory.Path, "cham.xsd");
+                string tempDirectoryPath = tempDirectory.Path[tempDirectory.Path.Length - 1] == Path.DirectorySeparatorChar ?
```
EnsureTrailingSlash

ViktorHofer replied:
This was just inlined to see if it works...

krwq replied:
This should not work on Windows. I've just tested this on Ubuntu and `ss.Add(null, XmlReader.Create(new StringReader(main), new XmlReaderSettings() { XmlResolver = new XmlUrlResolver() }, tempDirectoryPath));` is correctly resolving cham.xsd. Please file a bug and write a disabled test where you don't set it and you expect it to fail - I wouldn't fix it as part of this PR.

krwq replied:
last comment was supposed to be on different line - but is now more visible after I pushed to your branch :wink:

ViktorHofer replied:
I am confused now. With your latest commit, shouldn't the one test be passing now on e.g. Ubuntu? It still says all three tests failed

krwq replied:
you're right - it was passing on my ubuntu box and the CI I looked at. This is interesting test case. would be nice to have local repro

ViktorHofer replied:
If you find time it would be nice. Meanwhile I will try to test it on my Linux Subsystem.

krwq commented on src/System.Private.Xml/tests/XmlSchema/XmlSchemaSet/TC_SchemaSet_Compile.cs:
```diff
@@ -118,33 +118,42 @@ public void TFS_470021()
-                foreach (XmlSchemaObject child in seq.Items)
+                ss.Add(null, XmlReader.Create(new StringReader(cham)));
+                // TempDirectory path must end with a DirectorySeratorChar, otherwise it will throw in the Xml validation.
+                ss.Add(null, XmlReader.Create(new StringReader(main), null, tempDirectoryPath));
```
I think you're missing XmlUrlResolver

ViktorHofer replied:
But I'm not using XmlReaderSetting at all... Do I need to? And why only on Unix, OSX?

krwq commented:
@ViktorHofer I didn't notice there is more failures on some platforms - I don't see TFS_470021 failing at least - fix is going to be the same I believe - please file issues on that - would be good to know why there is a difference

ViktorHofer commented:
@sepidehMS I would like to forward this PR/Issue to you as I'm stuck. The idea is that three tests in System.Private.Xml tried to create a file in the appx folder which is not allowed inside an AppContainer. So I changed it to a TemporaryDirectory and added an overload to StringReader to pass the new path in. That worked for all Windows environments but fails in Unix and OSX. @krwq tried to solve it in one test and succeeded, but only for Ubuntu 14.04 and not for 16.04. 

It seems we have some compat issues. This could be a 2.0 production code issue... If you like we could follow up tomorrow.

cc @danmosemsft @karelz

sepidehMS commented:
@ViktorHofer I will investigate this.

sepidehMS commented:
After long investigation, I found out that the unexpected behavior in some configurations (Ubuntu14.04, OSX, CentOS, and PortableLinux) is due to XmlSchemaSet's dependency on `System.Uri.IsAbsoluteUri` when `UriKind=UriKind.AbsoluteOrRelative` and a behavior difference in System.Uri class . Opened https://github.com/dotnet/corefx/issues/20046 to track the issue.

danmosemsft commented:
@dotnet-bot test this please (get new logs)

danmosemsft commented:
As mentioned in #20046 it's expected the URI is relative on Unix because temp is relative  -- `/tmp/`.

sepidehMS commented:
Disabled the failing tests and created https://github.com/dotnet/corefx/issues/20148 to track the behavior difference in XmlSchemaSet.

ViktorHofer commented:
Great, thanks!
